drivers/video/ov7670.c

```diff
@@ -0,0 +1,482 @@
+		return -ENOTSUP;
+	}
+
+	if (!memcmp(&data->fmt, fmt, sizeof(data->fmt))) {
+		/* nothing to do */
+		return 0;
+	}
+
+	memcpy(&data->fmt, fmt, sizeof(data->fmt));
+
+	if (fmt->pixelformat == VIDEO_PIX_FMT_RGB565) {
+		com7 |= 0x4;
+	}
+
+	/* Set output resolution */
+	while (fmts[i].pixelformat) {
+		if (fmts[i].width_min == fmt->width && fmts[i].height_min == fmt->height &&
+		    fmts[i].pixelformat == fmt->pixelformat) {
+			/* Set output format */
+			switch (fmts[i].width_min) {
+			case 176: /* QCIF */
+				com7 |= BIT(3);
+				break;
+			case 320: /* QVGA */
+				com7 |= BIT(4);
+				break;
+			case 352: /* CIF */
+				com7 |= BIT(5);
+				break;
+			default: /* VGA */
+				break;
+			}
+			/* Program COM7 to set format */
+			return i2c_reg_write_byte_dt(&config->bus, OV7670_COM7, com7);
+		}
+		i++;
+	}
+	LOG_ERR("Unsupported format");
+	return -ENOTSUP;
+}
+
+static int ov7670_get_fmt(const struct device *dev, enum video_endpoint_id ep,
+			  struct video_format *fmt)
+{
+	struct ov7670_data *data = dev->data;
+
+	if (fmt == NULL) {
+		return -EINVAL;
+	}
+	memcpy(fmt, &data->fmt, sizeof(data->fmt));
+	return 0;
+}
+
+static int ov7670_init(const struct device *dev)
+{
+	const struct ov7670_config *config = dev->config;
+	int ret, i;
+	uint8_t pid;
+	struct video_format fmt;
+	const struct ov7670_reg *reg;
+
+	if (!i2c_is_ready_dt(&config->bus)) {
+		/* I2C device is not ready, return */
+		return -ENODEV;
+	}
+
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios)
+	/* Power up camera module */
+	if (gpio_is_ready_dt(&config->pwdn)) {
+		ret = gpio_pin_configure_dt(&config->pwdn, GPIO_OUTPUT_INACTIVE);
+		if (ret < 0) {
+			LOG_ERR("Could not clear power down pin: %d", ret);
+			return ret;
+		}
+	}
+#endif
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
+	/* Reset camera module */
+	if (gpio_is_ready_dt(&config->reset)) {
+		ret = gpio_pin_configure_dt(&config->reset, GPIO_OUTPUT);
+		if (ret < 0) {
+			LOG_ERR("Could not set reset pin: %d", ret);
+			return ret;
+		}
+		/* Reset is active low, has 1ms settling time*/
+		gpio_pin_set_dt(&config->reset, 0);
+		k_msleep(1);
+		gpio_pin_set_dt(&config->reset, 1);
+		k_msleep(1);
+	}
+#endif
+
+	/*
+	 * Read product ID from camera. This camera implements the SCCB,
+	 * spec- which *should* be I2C compatible, but in practice does
+	 * not seem to respond when I2C repeated start commands are used.
+	 * To work around this, use a write then a read to interface with
+	 * registers.
+	 */
+	uint8_t cmd = OV7670_PID;
+
+	ret = i2c_write_dt(&config->bus, &cmd, sizeof(cmd));
+	if (ret < 0) {
+		LOG_ERR("Could not request product ID: %d", ret);
+		return ret;
+	}
+	ret = i2c_read_dt(&config->bus, &pid, sizeof(pid));
+	if (ret < 0) {
+		LOG_ERR("Could not read product ID: %d", ret);
+		return ret;
+	}
+
+	if (pid != OV7670_PROD_ID) {
+		LOG_ERR("Incorrect product ID: 0x%02X", pid);
+		return -ENODEV;
+	}
+
+	/* Set default camera format (QVGA, YUV) */
```

Comment:
nit. `/* Set default camera format (QVGA, YUYV) */` is more precise as we have several "YUV" formats.